### DISCUSSION TITLE
Constellation plot using 3GPP channel models

### DISCUSSION THREAD

**Discussion by omsahoowork** [NVlabs/sionna#397] [Q&A] [answered]
I had a query. When I plot a constellation diagram using AWGN channel I was able to achieve so but when using any 3GPP channel model I am getting error.

Can anyone help me on how to get the constellation diagram plotted using any 3GPP channel model defined in sionna ????

**Comment by SebastianCa** [comment]
Hi @OmSahoo-Work,

Without further details about your experiment, it is difficult to offer specific advice. Could you please provide a code snippet that reproduces the error you're experiencing?

An example of a received (noisy) constellation diagram can be found in the [MIMO tutorial notebook](https://nvlabs.github.io/sionna/examples/Simple_MIMO_Simulation.html#Simple-uncoded-transmission). Please note that without equalization, the plots might not be very meaningful for such channels.

**Reply by omsahoowork** [reply]
Hi, I meant can we do a constellation plot like the below diagram using any of the 3gpp channel models cdl, uma, rma, etc...
If yes then how ?
![image](https://github.com/NVlabs/sionna/assets/152632519/8c8c2b69-ea85-4f71-98a3-98fda0a09794)

this is the error line, I am getting bits_hat from the neural receiver model trained using UMA channel.
I am passing it in mapper then want to obtain the above constellation diagram of bits_hat in the grid of imaginary and real symbols.

How to do so ?? is there any other way or there is any other kind of ofdm mapper class in sionna to obtain symbols ??
![image](https://github.com/NVlabs/sionna/assets/152632519/ceb99f2f-f545-4415-93f5-1a2b9c8f2ef2)

**Reply by SebastianCa** [reply]
Your post doesn't include an error message. Could you please provide a code snippet, ideally as a gist, that demonstrates the issue?

Have you considered using the approach from the [MIMO tutorial notebook](https://nvlabs.github.io/sionna/examples/Simple_MIMO_Simulation.html#Simple-uncoded-transmission) and substituting the `FlatFadingChannel` with your `UMa` channel? Additionally, instead of the `QAMSource`, you might try using a `Mapper` as demonstrated in the [Sionna tutorial notebook](https://nvlabs.github.io/sionna/examples/Sionna_tutorial_part1.html#Hello,-Sionna!).

One general note: I'm not entirely sure I've understood your experiment. However, if you hard-decide the output from a neural receiver and then re-encode/map the bits, your resulting constellation diagram does not show any channel effects.

**Reply by omsahoowork** [reply]
Hi, so I am using [this ](https://nvlabs.github.io/sionna/examples/Neural_Receiver.html) tutorial neural receiver and 
at the end as you can see we get bits_hat as output from the model. I want to see the bits_hat on a constellation diagram using the CDL channel as defined in the above notebook.

Can you help me how can I achieve that , get the constellation diagram using the cdl channel variations on bits_hat ?? 
I am simple doing this 
![image](https://github.com/NVlabs/sionna/assets/152632519/e1f7b9c1-2e32-48d4-9df1-411a89a46fbc)

Error
![image](https://github.com/NVlabs/sionna/assets/152632519/ecd46b13-49b7-4cf1-9ac5-b41130c4474d)

**Reply by SebastianCa** [reply]
In your case, the shape of `x` does not match the expected channel input. You should use a `ResourceGridMapper` to map `x` appropriately, as demonstrated for the baseline in the [tutorial notebook](https://nvlabs.github.io/sionna/examples/Neural_Receiver.html#End-to-end-System-as-a-Keras-Model).

Please note that the output `bits_hat` will contain a large number of random bits, specifically 128 * 768 * `num_bits_per_symbol`. When you plot these bits in a single constellation diagram (similar to a histogram), you will likely see a result comparable to using purely random bits instead of `bits_hat`. Thus, the insights of this setup appear limited to me.

**Reply by omsahoowork** [reply]
Then, can you suggest me how can I see the actual output bits of the neural receiver model on a constellation diagram using sionna ??

**Comment by SebastianCa** [comment, marked as answer]
Please understand that we cannot provide such detailed support.

It's not clear to me why you are looking to apply another `CDL` channel after decoding. 

If you aim to visualize the input of the neural receiver (i.e., the actual received symbols at the channel output), you will need to modify the end-to-end model such that it outputs `y`. This can then be displayed on a constellation diagram (as mentioned, this is only effective if you manually apply an equalizer).

If you wish to visualize `bits_hat`in a constellation diagram, you can simply apply the mapper and visualize the output.
